cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -306,4 +333,41 @@ private void publishStopProgram(String stopPayload) {
     programStateWriter.killed(programRunId);
     LOG.debug("Published program stop message for run {}", programRunInfo.getRun());
   }
+
+  /**
+   * Fetch program status updates and filter if they are initiated by a tethered peer.
+   */
+  private Map<String, TetheringProgramUpdate> getPeerProgramUpdates() {
+    String lastMessageId = lastProgramUpdateMessageId;
+    try {
+      Map<String, TetheringProgramUpdate> peerProgramUpdates = new HashMap<>();
+      try (CloseableIterator<Message> iterator = messageFetcher.fetch(NamespaceId.SYSTEM.getNamespace(),
+                                                                      programUpdateTopic,
+                                                                      programUpdateFetchSize,
+                                                                      lastProgramUpdateMessageId)) {
```

Comment:
where are we persisiting lastProgramUpdateMessageId so that we read from where we left off after restart?